Fix lambdify crash on empty tuple (#26119)

mohakmalviya opened the pull request:
Resolve Lambdify Crash on Empty Tuple Input
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #26119

#### Brief description of what is fixed or changed

Resolved a bug in lambdify where it crashed when handling an empty tuple. This was caused by a syntax error during the compilation of the function string. The fix involves modifying _recursive_to_string in sympy.utilities.lambdify to correctly handle empty tuples, ensuring they are converted to an empty tuple string in Python.
#### Other comments

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* utilities
  * Fix bug that caused lambdify to crash on empty tuples.
<!-- END RELEASE NOTES -->

mohakmalviya committed:
Fix lambdify crash on empty tuple (#26119)

Resolved a bug in lambdify where it crashed when handling an empty tuple. This was caused by a syntax error during the compilation of the function string.
The fix involves modifying _recursive_to_string in sympy.utilities.lambdify to correctly handle empty tuples, ensuring they are converted to an empty tuple string in Python.

Example of fixed behavior:
import sympy as sm
from sympy.printing.numpy import NumPyPrinter

a = sm.symbols("a")
sm.lambdify((a,), ((), (a,)), printer=NumPyPrinter) # Now works without SyntaxError

mohakmalviya committed:
author: add Senku to .mailmap

mohakmalviya committed:
author: add Senku to .mailmap

TJStienstra commented:
The solution looks good to me.

Could you add a simple test in [test_lambdify.py](https://github.com/sympy/sympy/blob/master/sympy/utilities/tests/test_lambdify.py)? The issue was by the way printer independent, so you do not need to specify the printer:
```python
a = sm.symbols("a")
sm.lambdify((a,), ((), (a,)))
```

sylee957 commented:
I also request changes for the release notes. Please follow the section

4. Write release notes in a way that can be understood by end-users of SymPy.

https://github.com/sympy/sympy/wiki/Writing-Release-Notes#what-makes-a-good-release-notes-entry

mohakmalviya committed:
added test for lambdify crashes on empty tuple

mohakmalviya commented:
> I also request changes for the release notes. Please follow the section
> 
> 4. Write release notes in a way that can be understood by end-users of SymPy.
> 
> https://github.com/sympy/sympy/wiki/Writing-Release-Notes#what-makes-a-good-release-notes-entry

thanks for sharing the link sir! i have edited my release notes according to the need

mohakmalviya commented:
> The solution looks good to me.
> 
> Could you add a simple test in [test_lambdify.py](https://github.com/sympy/sympy/blob/master/sympy/utilities/tests/test_lambdify.py)? The issue was by the way printer independent, so you do not need to specify the printer:
> 
> ```python
> a = sm.symbols("a")
> sm.lambdify((a,), ((), (a,)))
> ```

hello sir i have added the sample test please review it

tjstienstra commented:
> thanks for sharing the link sir! i have edited my release notes according to the need

The link mentions: _bug fixes and performance improvements do affect end-user behavior and should have release notes entries_

Could you therefore mention something like the following as a release note:
```
* utilities
  * Fix bug that caused lambdify to crash on empty tuples.
```

> hello sir i have added the sample test please review it

The added test looks good.

mohakmalviya commented:
> > thanks for sharing the link sir! i have edited my release notes according to the need
> 
> The link mentions: _bug fixes and performance improvements do affect end-user behavior and should have release notes entries_
> 
> Could you therefore mention something like the following as a release note:
> 
> ```
> * utilities
>   * Fix bug that caused lambdify to crash on empty tuples.
> ```
> 
> > hello sir i have added the sample test please review it
> 
> The added test looks good.

thankyou for correcting me sir! i have updated the release notes entry
